Reader profile - add support for a route via userId

Addison-Stavlo opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to # https://github.com/Automattic/loop/issues/369

## Proposed Changes

Allows use of a `/reader/users/id/{userId}` route that will then replace with `/reader/users/{username}` once the data is retrieved for the user.
* updates the action for `requestUser` to optionally request by id.
* creates a selector for the get-reader-user and uses this in the connect function of the user profile page.
* other related changes to accomadate this

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* So we can easily link to a reader profile from places where we don't easily have access to user_login but have the id.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Test the reader profile works as it did before.
* Verify there are no flashes of 'user not found' in the loading state.
* Verify going to /reader/users/id/{userId} directs to the correct user, loads properly, and replaces the url state in history with the updated login route value. (eg try /reader/users/5 vs. /reader/users/id/5)
* Verify there are no issues with trying inputs that will have an error response (e.g. /reader/users/id/foobar).
* Try to break it, test around.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

Addison-Stavlo committed:
support a route for /reader/users/id/:id

Addison-Stavlo committed:
update requestUser in reader state to handle finding by id

Addison-Stavlo committed:
create a selector for getting the reader user

Addison-Stavlo committed:
update userProfile to handle id route version

Addison-Stavlo committed:
fix loading bug regression

Addison-Stavlo committed:
fix bug with new selector iterating through a null value from unknown user response

Addison-Stavlo committed:
minor refactor on connect function

Addison-Stavlo committed:
whoopsie, dont rename that prop undo, undo

DustyReagan reviewed (approved):
This is working great for me!

I did catch a glimpse of 'user not found' in the loading state a couple of times by going directly to /reader/users/{username}, but I can't reproduce it consistently. In fact I can't reproduce it at all now. I tried clearing storage in web dev tools and testing in a private browser, but I still can't reproduce it. I'd say it's not a blocker, but it also might still need a follow-up.

Addison-Stavlo commented:
>  tried clearing storage in web dev tools and testing in a private browser, but I still can't reproduce it. I'd say it's not a blocker, but it also might still need a follow-up.

Ok! Lets keep an eye out. I saw it on an earlier commit when I renamed one of the action props incorrectly, but stopped seeing it after i fixed that.
